### ISSUE TITLE
SEGFAULT when decoding numpy ndarray of dtype object

### ISSUE THREAD

**Issue by rfdougherty** [jsonpickle/jsonpickle#226]
Decoding an ndarray of dtype object that's large enough to trigger binary encoding will result in a segfault. E.g.,

```
import jsonpickle
import jsonpickle.ext.numpy as jsonpickle_numpy
jsonpickle_numpy.register_handlers()
import numpy as np

arr = np.array(['F'+str(i) for i in range(30)], dtype=np.object)
buf = jsonpickle.encode(arr)
del arr  # or save buf to disk, quit, restart, and attempt to load from disk
print(jsonpickle.decode(buf))

```

This seems to be a bug in numpy when restoring a binary-encoded ndarray of dtype object. E.g., this snippet will segfault in the same way:

```
    import numpy as np
    arr = np.array([str(i) for i in range(3)], dtype=np.object)
    dtype = arr.dtype
    shape = arr.shape
    buf = arr.tobytes()
    del arr
    arr = np.ndarray(buffer=buf, dtype=dtype, shape=shape).copy()
```

A work-around is to catch ndarrays of dtype object and encode them as a list.

**Comment by davvid** [comment]
@jaraco @shakhat @EelcoHoogendoorn @guilhemchalancon @aldanor do you have any opinions on how to handle this topic?

One the one hand I am sympathetic to users that don't like to see segfaults.  On the other hand, we're just a canary in the coalmine; surely the numpy developers have already fixed this by now?  Special-casing the broken numpy version(s) seems like the way to go if we decide to workaround this bug, despite the cost in code complexity.  Thoughts?

IMO, ideally we could just close this issue and users (distros included) will avoid the broken numpy versions, but I don't know the particulars of this issues.  @rfdougherty do you know if this has been reported to the numpy developers?  This looks like a reference counting bug.

**Comment by jaraco** [comment]
I don't use numpy, but given the 'ext' packaging and the fact that the issue occurs upstream without any jsonpickle involvement, I'm inclined to agree with @davvid - let's let the users who wish to use numpy and jsonpickle find/develop unbroken versions to use to avoid the bug. I definitely don't think jsonpickle should take on the responsibility of working around upstream bugs. What I recommend is that the pure numpy behavior be reported upstream to assess that it is an actual bug.

> why not just call .tolist()

Often when you're serializing an object or graph of objects, it's not readily possible to transform the input objects readily. In addition, you get better fidelity if the object you deserialize matches the interface of the object you serialized. If you require the caller to `tolist()` on serialization, they'll need to `toarray()` on the other end. It's this kind of logic that jsonpickle attempts to abstract. IMHO, if jsonpickle can reliably serialize and deserialize numpy objects (in the absence of numpy bugs), that's a reasonable feature to support.

**Comment by EelcoHoogendoorn** [comment]
I agree of course; the deserialized datastructure should ideally be indistinguishable from the original; and I dont think its technically impossible; but I really dont see it being very easy. There are various things id like to improve about jsonpickle_numpy; but this feature is a few orders of magnitude down the list in terms of effort/reward.

Looking at the segfaulting numpy snippet, it is not even a bug though, and I wouldnt waste my breath waiting for that behavior to change upstream. It is simply an unsafe operation to do binary manipulations on what is essentially an array of pointers. This isnt a numpy bug; this is just the essence of why implementing pickle-style logic, while interfacing with a C-library that has its own memory management logic independent from python's reference counting scheme, is hard.

So until someone implements a piece of logic that passes a unit test where a roundtrip of serialization leaves an object-array unchanged (it could be a while), throwing an exception is clearly preferable to a segfault. And adding a flag to allow for implicit casting of objects arrays to lists removes the need for manual transforms. Not on both sides, but thats ok if it does not happen silently without you opting into it id say.

**Comment by EelcoHoogendoorn** [comment]
I see some work has been done on NumpyNDArrayHandlerBinary already; a few notes:

- Adding the tolist() case could be implemented more elegantly by defering to the superclass; it handles a few edge-quirks like arrays with shapes containing zeros, for instance; which do not deserialize correctly otherwise.
- Also, is the binary compression on the result of tolist() safe, the way it is implemented now? What if the ndarray references objects that are also present elsewhere in the serialized datastructure? Would that resolve correctly? Not sure, but probably good to test. Deferring to the base class could solve that too.
- Id really like that flag to configure the response to this behavior; analogous to the flag one can pass to the view-serializing class. I think the default ought to be an exception.
